### PR TITLE
Added support for non-square slices.

### PR DESCRIPTION
**Pull request by bwvdnbro**
As discussed in #115, support for non-square images was lacking from `visualisation/slice.py`.

I have provided a minimal form of support that avoids some of the potential pitfalls: instead of normalising the pixel coordinates using the range in both the x and y dimension, the maximum of these ranges is used. A square image with the requested resolution is then produced that will contain empty pixels in the smaller dimension. These empty pixels are filtered out before returning the non-square result image.

This minimal implementation implies that
 - the image resolution argument is applied to the longer dimension only (after possible rotations) and should be interpreted as such
 - there could be a small mismatch in the pixel length of the smaller dimension because of rounding down
 - no stretching happens, because all coordinates are normalised with the same number

### CODE REVIEW

**Comment by JBorrow** [comment]
Please add a relevant test and can you show an example image (e.g. the blob test?)

**Comment by bwvdnbro** [comment]
I have added a test.

This is what it looks like when applied to the output of the 3D Sod shock:
gas_slice_map

Script:
```
from swiftsimio import load
from swiftsimio.visualisation.slice import slice_gas
from swiftsimio.visualisation.rotation import rotation_matrix_from_vector

data = load("sodShock_0001.hdf5")

# This creates a grid that has units msun / Mpc^3, and can be transformed like
# any other unyt quantity. Note that `slice` is given in terms of the box-size,
# so here we are taking a slice at z = boxsize / 2.
mass_map = slice_gas(
    data,
    slice=0.5,
    resolution=1024,
    project="masses",
    parallel=True,
)

# Let's say we wish to save it as g / cm^2,
from unyt import g, cm

mass_map.convert_to_units(g / cm**3)

from matplotlib.pyplot import imsave
from matplotlib.colors import LogNorm

# Normalize and save
imsave("gas_slice_map.png", LogNorm()(mass_map.value), cmap="viridis")
```

**Comment by JBorrow** [comment]
Final question is are you sure this doesn't affect the units in slice_gas?

**Comment by bwvdnbro** [comment]
That question opened a whole can of worms...

Short answer: no.

Long answer: the units are wrong and I don't think the slicing function works correctly if `x_range` and `y_range` are not equal to the box size in `z`, even in the old version. The reason is that `slice_scatter` assumes that the re-scaling from coordinates to `[0,1]` is shape-preserving (because the re-scaled kernel should still be spherical), which really only works if the denominator in the re-scaling is the same for all 3 coordinate directions. The old version only guaranteed `x_range==y_range` and completely ignored `z`. That could explain why the Sod shock image looks so noisy (while the Sod curves are reasonably well behaved).

This would be pretty easy to fix if there were no rotations. With rotations, the length in `z` is ill-defined and I don't really know how to fix this.

**Comment by bwvdnbro** [comment]
I just pushed a new version that fixes the units, but will not work properly with rotations. @JBorrow I would really like your input on this to decide what to do about the `z` direction.

Updated Sod shock image:
gas_slice_map

**Comment by JBorrow** [comment]
Well to me it seems like `z` should have no bearing on the output _units_, as they're 3D? But I certainly see the issue with rotations.

I think we could make this a physical offset from the centre (which is x,y,z = 0.5 * boxsize by default)?

**Comment by bwvdnbro** [comment]
I agree that it's not really the _units_ per se. However, because of the re-scaling, the map is always produced in units of [Q]/(Lx Ly Lz), where Lx, Ly, Lz are the box size or region size in the different directions and [Q] is the unit of the mapped quantity. What I'm arguing here is that Lx, Ly and Lz should always be the same to conserve shapes when you do the re-scaling (e.g. the shape of the _spherical_ smoothing kernel). Right now, Lz is always chosen to just be the box size in z, without any restrictions on its value. While it really should be Lz=L, with L = max(Lx, Ly). I have changed that in the last commit.

**Comment by JBorrow** [comment]
I'm not sure that's the case. Here we only use `z` and `h` to compute the distance from the plane... So as long as the `z` units and `h` units are the same, you're good to go, no?

**Comment by bwvdnbro** [comment]
Yes. So the `z` and `h` units have to be the same to check the overlap with the slice, and the `x`, `y` and `h` units also have to be the same for the kernel sampling to work. Logic then dictates that `x`, `y` and `z` have to have the same units, so even though you don't care about `z`, you still need to make sure the re-sampling of `x` and `y` conserves 3D shapes.

The only difference between the slice re-sampling and the projection re-sampling is that you don't care about the re-sampling factor for `z`. So the box size in `z` does not enter the equations. Is that what you mean?

**Comment by bwvdnbro** [comment]
Since @Fonotec has no found an example that returns the wrong density in a slice (because of the wrong scale factor for the z coordinates), I had another go at cleaning this up. I have now changed the meaning (and name) of the `slice` argument to the functions so that it is a physical offset rather than some fraction of the box size. I thought a sensible default would be the origin of the box, since that seems more natural than the center if you use particle coordinates to pick the location of the slice. If a rotation center is provided, that is used instead, since you probably want to put your slice close to the center of an object you just rotated. We can argue about those choices.

I had to update the unit tests to reflect these changes, and I suspect the documentation also needs to be updated.

I think we might need to release a new version once these changes are made, since the slicing functionality is effectively broken if you do not use the full (cubic) box.

File: swiftsimio/visualisation/slice.py [hunk 1]
```diff
@@ -396,13 +401,18 @@ def slice_gas_pixel_grid(
...
+
+    # trim the image to remove empty pixels
+    return image[:xres, :yres]
 
 
 def slice_gas(
     data: SWIFTDataset,
     resolution: int,
-    slice: float,
+    z_slice: unyt_quantity = 0.0 * unyt.m,
```
Comment (JBorrow) [inline swiftsimio/visualisation/slice.py:415]: I don't like this - can we have `z_slice: Optional[unyt_quantity] = None` and set it to `0.0 * data.gas.coordinates.units`?

**Comment by JBorrow** [comment]
I'm happy to merge this after there have been associated documentation improvements.

**Comment by bwvdnbro** [comment]
I think this should be ready to go now.

**Pull request merged**